metrics_utility/test/test_db_lock_helpers.py

```diff
@@ -0,0 +1,115 @@
+from concurrent.futures import ThreadPoolExecutor
+from unittest.mock import MagicMock, patch
+
+import pytest
+
+from metrics_utility.automation_controller_billing.collector import Collector
+from metrics_utility.base.db_lock_helpers import advisory_lock
+
+
+class TestCollectorLocks:
+    def acquire_lock(self, thread_id, key='contention_test_lock'):
+        try:
+            with advisory_lock(f'{key}', False) as acquired:
+                if acquired:
+                    lock = f'Thread {thread_id} acquired lock. Key: {key}'
+                    return lock
+                else:
+                    lock = f'Thread {thread_id} did not, and should not, acquire lock. Key: {key}'
+                    return lock
+        except Exception as e:
+            lock = f'Thread {thread_id} encountered error: {e}. Key: {key}'
+            return lock
```

Comment:
please don't use threads if we don't have to

this could be as simple as 

```suggestion
    def test_acquire_lock(self):
        with advisory_lock('test', False) as acquired:
            assert acquired is not None
            with pytest.raises(Exception) as excinfo:
                with advisory_lock('test', False) as acquired2:
                    assert False, "this should be unreachable"
```

feel free to remove `test_collector_lock_behavior` & `test_gahter_calls_lock` - in the library, collectors don't take locks, the task does

does `test_string_key_conversion` do anything anymore? since this is always a string?

the rest, up to you, but no `ThreadPoolExecutor` please :)

Reply:
One thing that needs to be considered is also what is the scope of the lock - I believe its database session. This means, if there will be one DB session, second inner lock call will actually pass (this is called reentrant locking, its common and chatgpt tells me that this is the case of postgres lock, we may to double check this).

If this is true, the suggestion from Martin will not throw exception and will not work as intended, in this case:

We must make sure that we are using different DB sessions for each lock.

Simple theory behind:
Every lock has owner, if the owner has lock and wants it to acquire the lock again, it will succeed. There is lock counter that will count the count of acquire and every unlock will lower this count. So:

lock
      lock (will pass for same session, now locked for second time)
      unlock (we are still locked!!)
unlock (now we are finally unlocked)

So Im not sure if we will be able to do it without threads :-( Because if I understand it correctly, DB connections are reused for thread - so each thread holds single session - thus if you will use lock twice inside one thread (DB session), it will acquire lock without problems. But if there is simple way how to use two sessions and pass them to acquire lock, this would work...

If not, I suggest leave it be and approve it (unless we come up with simple solution), locking is not crucial for our app anyway.

Reply:
Nah, no worries, it still works :) 

The db requests are all serialized through one thread anyway.

Tested with...

```python
import pytest

from django.db import connection

from metrics_utility.library import lock


def test_acquire_lock():
    with lock(connection, 'abc', False) as acquired:
        assert acquired is not None
        with pytest.raises(Exception):
            with lock(connection, 'abc', False):
                assert False, 'unreachable'
```